src/ray/common/function_descriptor.h

```diff
@@ -0,0 +1,181 @@
+  }
+
+  virtual std::string ToString() const {
+    return "{type=JavaFunctionDescriptor, class_name=" + typed_message_->class_name() +
+           ", function_name=" + typed_message_->function_name() +
+           ", signature=" + typed_message_->signature() + "}";
+  }
+
+  std::string ClassName() const { return typed_message_->class_name(); }
+
+  std::string FunctionName() const { return typed_message_->function_name(); }
+
+  std::string Signature() const { return typed_message_->signature(); }
+
+ private:
+  const rpc::JavaFunctionDescriptor *typed_message_;
+};
+
+class PythonFunctionDescriptor : public FunctionDescriptorInterface {
+ public:
+  /// Construct from a protobuf message object.
+  /// The input message will be **copied** into this object.
+  ///
+  /// \param message The protobuf message.
+  explicit PythonFunctionDescriptor(rpc::FunctionDescriptor message)
+      : FunctionDescriptorInterface(std::move(message)) {
+    RAY_CHECK(message_->function_descriptor_case() ==
+              ray::FunctionDescriptorType::kPythonFunctionDescriptor);
+    typed_message_ = &(message_->python_function_descriptor());
+  }
+
+  virtual size_t Hash() const {
+    return std::hash<int>()(ray::FunctionDescriptorType::kPythonFunctionDescriptor) ^
+           std::hash<std::string>()(typed_message_->module_name()) ^
+           std::hash<std::string>()(typed_message_->class_name()) ^
+           std::hash<std::string>()(typed_message_->function_name()) ^
+           std::hash<std::string>()(typed_message_->function_hash());
+  }
+
+  virtual std::string ToString() const {
+    return "{type=PythonFunctionDescriptor, module_name=" +
+           typed_message_->module_name() +
+           ", class_name=" + typed_message_->class_name() +
+           ", function_name=" + typed_message_->function_name() +
+           ", function_hash=" + typed_message_->function_hash() + "}";
+  }
+
+  std::string ModuleName() const { return typed_message_->module_name(); }
+
+  std::string ClassName() const { return typed_message_->class_name(); }
+
+  std::string FunctionName() const { return typed_message_->function_name(); }
+
+  std::string FunctionHash() const { return typed_message_->function_hash(); }
+
+ private:
+  const rpc::PythonFunctionDescriptor *typed_message_;
+};
+
+typedef std::shared_ptr<FunctionDescriptorInterface> FunctionDescriptor;
+
+inline bool operator==(const FunctionDescriptor &left, const FunctionDescriptor &right) {
+  if (left.get() != nullptr && right.get() != nullptr && left->Type() == right->Type() &&
+      left->ToString() == right->ToString()) {
+    return true;
+  }
+  return left.get() == right.get();
+}
+
+inline bool operator!=(const FunctionDescriptor &left, const FunctionDescriptor &right) {
+  return !(left == right);
+}
+
+/// Helper class for building a `FunctionDescriptor` object.
+class FunctionDescriptorBuilder {
+ public:
+  /// Build a DriverFunctionDescriptor.
+  ///
+  /// @return new ray::FunctionDescriptor
```

Comment:
```suggestion
  /// \return new ray::FunctionDescriptor
```
We use this style elsewhere